openssl/src/ts.rs

```diff
@@ -0,0 +1,457 @@
+//! Partial interface to OpenSSL Time-Stamp Protocol (RFC 3161) implementation.
+//!
+//! This module provides a partial interface to OpenSSL's TSP implementation.
+//! The aim is to provide enough functionality for a client to request and
+//! verify timestamps returned by a Time Stamp Authority.
+use foreign_types::{ForeignType, ForeignTypeRef};
+use libc::{c_int, c_long, c_uint};
+
+use std::ptr;
+
+use asn1::{Asn1IntegerRef, Asn1ObjectRef};
+use error::ErrorStack;
+use ffi;
+use hash::MessageDigest;
+use pkey::{HasPrivate, PKeyRef};
+use x509::{X509Algorithm, X509Ref};
+use {cvt, cvt_p};
+
+foreign_type_and_impl_send_sync! {
+    type CType = ffi::TS_MSG_IMPRINT;
+    fn drop = ffi::TS_MSG_IMPRINT_free;
+
+    /// A message imprint contains the hash of the data to be timestamped.
+    pub struct TsMsgImprint;
+
+    /// Reference to `TsMsgImprint`.
+    pub struct TsMsgImprintRef;
+}
+
+/// A builder used to construct a `TsMsgImprint`
+pub struct TsMsgImprintBuilder(TsMsgImprint);
+
+impl TsMsgImprintBuilder {
+    /// Creates a new builder.
+    pub fn new() -> Result<TsMsgImprintBuilder, ErrorStack> {
+        unsafe {
+            ffi::init();
+            let imprint: *mut ffi::TS_MSG_IMPRINT = cvt_p(ffi::TS_MSG_IMPRINT_new())?;
+            Ok(TsMsgImprintBuilder(TsMsgImprint::from_ptr(imprint)))
+        }
+    }
+
+    /// Sets the algorithm identifier of the message digest algorithm.
+    ///
+    /// This corresponds to `TS_MSG_IMPRINT_set_algo`.
+    pub fn set_algo(&mut self, digest: &MessageDigest) -> Result<(), ErrorStack> {
```

Comment:
It seems like this should take an X509Algorithm rather than a MessageDigest.

Reply:
Hmm, in that case I think it should be possible to easily construct an X509Algorithm from a MessageDigest. Now there's no public interface for constructing an X509Algorithm at all.
Also, according to RFC 3161 only message digest algorithms are valid here: "The hash algorithm indicated in the hashAlgorithm field SHOULD be a known hash algorithm (one-way and collision resistant)." -- that's why I thought that using a MessageDigest on this interface would be more precise and useful.

Reply:
Oh, and OpenSSL is inconsistent here: TS_MSG_IMPRINT_set_algo() uses X509_ALGOR, but its counterpart on the server side, TS_RESP_CTX_add_md() takes an EVP_MD. The server-side only accepts imprints with digest algorithms explicitly added via TS_RESP_CTX_add_md().